torch/utils/hooks.py

```diff
@@ -14,14 +14,21 @@ class RemovableHandle:
         hooks_dict (dict): A dictionary of hooks, indexed by hook ``id``.
         extra_dict (dict): An additional dictionary whose keys will be deleted
             when the same keys are removed from ``hooks_dict``.
+        module (nn.Module): If passed, the hook dict corresponds to that module,
+            otherwise it is a global hook dict.
     """
 
     id: int
     next_id: int = 0
 
-    def __init__(self, hooks_dict: Any, *, extra_dict: Any = None) -> None:
+    def __init__(self, hooks_dict: Any, *, extra_dict: Any = None, module=None) -> None:
         self.hooks_dict_ref = weakref.ref(hooks_dict)
         self.id = RemovableHandle.next_id
+
+        # TODO: we don't pickle/unpickle this field, which means the 'update_has_hooks'
+        # functionality (which is an optimization) decays after pickling.  Can we fix this?
```

Comment:
It is NOT an optimization, you rely on it to be correct to have the right behavior! That sounds bad?

Reply:
Here's why i thought of it as an optimization (maybe i missed a case?)

1) you save a TS module that _has_ hooks and you pickle a RemovableHandle associated with a hook
2) load the scriptmodule and the handle
3a) run the module: the module correctly recomputes the value of '_has_hooks' on load
3b) remove the hook using the unpickled handle and then run the module:  unclear if remove even works by itself, after save/load? i doubt it.  But regardless, if the module still thinks it '_has_hooks' and the hook is removed, all that happens is the fast path will be skipped and the slow path will check the hooks dicts